daemon/runtime_unix.go

```diff
@@ -134,3 +137,35 @@ func (daemon *Daemon) getRuntime(name string) (*types.Runtime, error) {
 
 	return rt, nil
 }
+
+// isValidC8dRuntimeName tests whether name
+// is a well-formed ContainerD runtime name,
+// with additional Docker-specific restrictions.
+//
+// A runtime name containing slash characters is interpreted by ContainerD
+// as the path to a runtime binary.
+// If we allowed this, anyone with Engine API access could get ContainerD
+// to execute an arbitrary binary as root.
+// Although Engine API access is already equivalent to root on the host,
+// the runtime name has not historically been a vector
+// to run arbitrary code as root
+// so users are not expecting it to become one.
+//
+// This restriction is not configurable.
+// There are viable workarounds for legitimate use cases:
+// administrators and runtime developers can make runtimes available
+// for use with Docker by installing them onto $PATH
+// following the [binary naming convention] for ContainerD Runtime v2.
+//
+// [binary naming convention]: https://github.com/containerd/containerd/blob/main/runtime/v2/README.md#binary-naming
+func isValidC8dRuntimeName(name string) bool {
+	// ContainerD uses a rather permissive test to validate runtime names.
+	// Any name which meets either requirement is valid:
+	//   - filepath.IsAbs(name), or
+	//   - the name contains at least one '.' character and no '/' characters.
+	// The name "." is technically valid per ContainerD's implementation:
+	// it would resolve to a binary named "containerd-shim---".
+	// https://github.com/containerd/containerd/blob/11ded166c15f92450958078cd13c6d87131ec563/runtime/v2/manager.go#L297-L317
+	// https://github.com/containerd/containerd/blob/11ded166c15f92450958078cd13c6d87131ec563/runtime/v2/shim/util.go#L83-L93
+	return !filepath.IsAbs(name) && !strings.ContainsRune(name, '/') && strings.ContainsRune(name, '.')
```

Comment:
This path would always be "local", for containerd, right? (i.e., no need to take cross-platform into account) otherwise we would need something like https://github.com/docker/cli/pull/1990


I think ideally;

- we should consume some function from containerd that performs the validation, so that we don't diverge
- perhaps create a tracking issue to contribute such a function in containerd (i.e. [based on this code](https://github.com/containerd/containerd/blob/11ded166c15f92450958078cd13c6d87131ec563/runtime/v2/manager.go#L297-L317)
- wondering if the containerd validation is (per the above) _too_ relaxed, i.e., does it want to allow `con....tai-----nerd-🤣🐳-shim...---` as name? (something like [`RestrictedNameChars`](https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/daemon/names/names.go#L6), possibly tweaked to not allow consecutive separators may be good?